src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/DetectPLINQNopsAnalyzer.cs

```diff
@@ -0,0 +1,115 @@
+
+        internal static readonly DiagnosticDescriptor DefaultRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             localizableTitle,
+                                                                             s_localizableMessageDefault,
+                                                                             DiagnosticCategory.Usage,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             description: s_localizableDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(ctx =>
+            {
+                if (!ctx.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqParallelEnumerable, out var parallelEnumerable))
+                {
+                    return;
+                }
+
+                if (!ctx.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var linqEnumerable))
+                {
+                    return;
+                }
+
+                var asParallelSymbols = parallelEnumerable.GetMembers("AsParallel").ToImmutableHashSet();
+                var collectionSymbols = parallelEnumerable.GetMembers("ToArray")
+                .Concat(parallelEnumerable.GetMembers("ToList"))
+                .Concat(parallelEnumerable.GetMembers("ToDictionary"))
+                .Concat(linqEnumerable.GetMembers("ToHashSet"))
+                .ToImmutableHashSet();
+
+                ctx.RegisterOperationAction(x => AnalyzeOperation(x, asParallelSymbols, collectionSymbols), OperationKind.Invocation);
+            });
+        }
+
+        public static bool ParentIsForEachStatement(IInvocationOperation operation) => operation.Parent is IForEachLoopOperation || operation.Parent?.Parent is IForEachLoopOperation;
+
+        public static bool TryGetParentIsToCollection(IInvocationOperation operation, ImmutableHashSet<ISymbol> collectionSymbols, out IInvocationOperation parentInvocation)
+        {
+            parentInvocation = operation;
+            if (operation.Parent?.Parent is not IInvocationOperation invocation)
+            {
+                if (operation.Parent?.Parent?.Parent is IInvocationOperation nestedInvocation) // Happens on ToHashSet calls
```

Comment:
I'll let @mavasani give some insight here but it feels weird to have this dependency on the operation structure.

Reply:
It was indeed weird, this is my first time writing an IOperation analyzer so it is quite possible I did miss some other functionality to achieve this check

Reply:
I guess it could also be expressed as FindFirstIInvocationMatchingMethodSymbol

Reply:
Make sure that you bail out on seeing a lambda or local function parent.